Printer to print local children of imported components

kerimoyle opened the pull request:
Addresses #590.

Additional to this test, I'd also expect to be able to specify this relationship (equivalence between imported variables) through the API.  This isn't possible at the moment?

Also closes #609

kerimoyle committed:
Test showing printer missing equivalent variables

kerimoyle committed:
Second test to show what I'd want to be able to do ...

hsorby commented on tests/variable/variable.cpp:
```diff
@@ -1699,3 +1699,38 @@ TEST(Variable, variableInterfaceDontDowngrade)
+    // Create equivalent variables: THIS BIT I DON'T KNOW HOW TO DO!
+    // The importedComponent is not instantiated, so can't point to any of its variables...
+    // libcellml::Variable::addEquivalence(x, importedComponent->variable("x"));
+    libcellml::Variable::addEquivalence(localComponent, "x", importedComponent, "x");
```
Why aren't you using the reference to 'x' that you have for the local component?  If you believe that imported component also has a variable 'x' you can create that variable and add it to the component.  With that you can use the API.

I could see us having the API (component, variable, component, string)

kerimoyle replied:
And what about when both components are imported? Then you'd need (component, string, component, string)

kerimoyle replied:
And if I add a variable to the importedComponent, isn't that a bit weird? Then I have two variables with the same name?

hsorby replied:
Having two variables with the same name isn't weird, there are loads of duplicate variable names.

hsorby replied:
I think then that you would make the whole set:

- (component, variable, component, string)
- (component, string, component, string)
- (component, string, component, variable)
- (component, string, component, string)

kerimoyle replied:
Duplicate variable names, sure, but not in the same component?
But yes to the whole set of combinations, though you have a duplicate in there ... :)

kerimoyle replied:
or, better:
variable, variable
variable, component, string,
(possibly?) component, string, variable (?)
component, string, component, string

nickerso replied:
I'd thought last time that I'd talked with @hsorby about this issue, the idea was that you would simply add variables to the imported component and then the equivalences work as normal. But then when printing the model the imported component would not have the variables printed (as they make no sense) and but the equivalences (connections) would still be printed.

hsorby replied:
Yes but I can't see any reason why you couldn't add the string form for convenience.

This is possible at the moment but you can only do it one way. I think if adding a string form would make it easier for users and doesn't make problems and tests still pass at the moment I cannot say it is wrong to add the additional API.

kerimoyle replied:
@nickerso I don't quite follow the logic of adding the variable to the component when it's really already there in the import?  Do you also have to add units for the model (before flattening) to pass validation?  Do you have to also have the units in your instantiated model too?  I'm missing something ...

nickerso replied:
> Yes but I can't see any reason why you couldn't add the string form for convenience.

Yep, totally agree.

nickerso replied:
> it's really already there in the import

Its only there once the import is resolved and instantiated. For exactly this case of handling CellML connections as variable equivalences it is sometimes necessary to make these dummy variables. They are never used for anything else (I think) so units/interfaces/etc. are not needed.

nickerso replied:
and since this would be an imported component these dummy variables are never used in flattening, validation, printing...

kerimoyle replied:
The units are used in validation because we now test for unit matching across equivalent variables.  :( I get an error:
```Validator error[0]:
     Description: Variable 'g_Na' in component 'sodiumChannel' has units of 'mS_per_cm2' and an equivalent variable 'g_Na' in component 'sodiumChannelInit' with non-matching units of ''. The mismatch is: 
    See section X.Y.Z in the CellML specification.```

... where `sodiumChannel` is a local component, and `sodiumChannelInit` is an import.  Is this expected? Or should the validator not test for units in a component which has a non-null import source?

kerimoyle replied:
Also interface errors from the dummy variables: 
```
Description: Variable 'm' in component 'mGateInit' has no interface type set. The interface type required is 'public'.
```

hsorby commented:
To be clear it is possible to do this at the moment.

kerimoyle commented:
So is this a thing? What should I do in the tutorials? Use dummy variables or wait for an overloaded function?

nickerso commented:
> What should I do in the tutorials? Use dummy variables or wait for an overloaded function?

Either push ahead with the dummy variable method in the tutorial (this works now) or implement the proposed new API for the convenience methods and use them in the tutorials.

Personally, I'm not sure how much of a convenience that these methods would be so I'd continue on the tutorials with the dummy variables.

kerimoyle committed:
Merge branch 'develop' into i590_print_model_with_imports

kerimoyle commented:
I'm going to revert the second commit, as discussion about workarounds there has obscured the actual bug ... I'll add the other as a "nice to have maybe" issue on its own.

kerimoyle committed:
Revert "Second test to show what I'd want to be able to do ..."

This reverts commit 9c701e0a752c1b9fc4e77eda3670d9b5601a2492.

kerimoyle committed:
Merge branch 'i590_print_model_with_imports' of https://github.com/kerimoyle/libcellml into i590_print_model_with_imports

kerimoyle committed:
Added more tests to show encapsulations with imports not parsed

kerimoyle committed:
bug in test resources

kerimoyle committed:
and again ...

kerimoyle committed:
Update file_parser.cpp

kerimoyle committed:
moved test to printer as that's where the trouble is ...

kerimoyle committed:
fixed order of components in resources file

kerimoyle committed:
Removed validation check from parser test (bug there addressed in #605)

kerimoyle committed:
Fixed bug in printer

kerimoyle commented:
Wow, this PR went all over the show!  In the end the bug was in the `Printer`, not the `Parser` as I'd originally thought.  This PR has three aspects:
- Fixed the bug in the printer that was missing local children of imported components (src/printer.cpp).  This is the main bit!
- Improved the debug printing of the model structure (test_utils.cpp)
- Added tests for printing/parsing of imports, encapsulations, and connections (tests and resources files)

Ready for review please @agarny @hsorby @nickerso

kerimoyle committed:
Merge branch 'develop' into i590_print_model_with_imports

kerimoyle committed:
Removed debug statements

kerimoyle committed:
Removed comments

kerimoyle committed:
Merge branch 'develop' into i590_print_model_with_imports

kerimoyle committed:
removed overloads in test utils printing functions

kerimoyle committed:
Merge branch 'develop' into i590_print_model_with_imports

kerimoyle committed:
Merge branch 'develop' into i590_print_model_with_imports

kerimoyle commented:
Ping @agarny @hsorby @nickerso

kerimoyle committed:
Merge branch 'develop' into i590_print_model_with_imports

kerimoyle committed:
Merge branch 'develop' into i590_print_model_with_imports

hsorby committed:
Merge in changes from 'prime/develop' fixing conflicts.

hsorby committed:
Change name of print model and compnent test utility functions.

agarny committed:
Test utils: reworked the methods for pretty printing a model/component.

agarny committed:
Printer: slight improvement to the code.

agarny committed:
Tests: slight speed improvements.

agarny committed:
Removed unneded test CellML files.

agarny committed:
Some minor cleaning up.

agarny reviewed (dismissed):
Looks good. I have cleaned up various things though. One thing that I would like to see however is for the `connectionsPersistAfterImporting`, `connectionsAsymmetricWithRepeatedUnits` and `connectionsSymmetricWithMismatchedUnits` to check the serialisation of the model. Right now, we test that this or that component/variable exists, etc. which makes the test rather difficult to read/understand (for me, at least).

hsorby commented on tests/test_utils.cpp:
```diff
@@ -60,44 +60,27 @@ void printIssues(const libcellml::LoggerPtr &l, bool headings, bool causes, bool
-}
-
-void prettyPrintComponentToStdout(const libcellml::ComponentPtr &component, size_t const c, std::string const spacer, bool includeMaths)
+void printComponent(const libcellml::ComponentPtr &component, size_t c, const std::string &indent, bool includeMaths)
```
This was recently changed to 'prettyPrint*' are we done with that now?

agarny replied:
The utility was originally named `printComponent()` and I don't see why it was renamed to something unnecessarily long and that I would never remember, hence I reverted the name of the method to what it is in `develop`.

hsorby commented on tests/test_utils.cpp:
```diff
@@ -60,44 +60,27 @@ void printIssues(const libcellml::LoggerPtr &l, bool headings, bool causes, bool
 
-    std::cout << spacer << local << "VARIABLES: " << component->variableCount() << " variables" << std::endl;
+    std::cout << std::endl;
+    std::cout << indent << INDENT << "VARIABLES: " << component->variableCount() << " variables" << std::endl;
```
Mixing use of 'indent' and 'INDENT' here and in other places.

agarny replied:
No, no mixing. `INDENT` is a constant used to indent while `indent` is an indent that we need for calling `printComponent()` recursively.

hsorby commented on tests/printer/printer.cpp:
```diff
@@ -308,13 +309,14 @@ TEST(Printer, printModelImportingModelParentComponent)
     auto serialisedModel = printer->printModel(model);
 
-    EXPECT_EQ(serialisedModel, fileContents("importingModelParentComponent.cellml"));
+    EXPECT_EQ(serialisedModel, modelContents);
```
These entries should be the other way around, we expect 'modelContents' to be equal to what we actually got 'serialiseModel' after our manipulations.

agarny replied:
Good point. Done in commit 2af7803c.

hsorby reviewed (dismissed):
A few minor things.  Importantly why change the name of the test utility printing functions again?

agarny committed:
Tests: put the expected/actual bit in the right order.

kerimoyle commented:
I can't quite tell who can approve this now that three of us have worked on it? @hsorby @agarny @nickerso ?

hsorby committed:
Merge branch 'develop' into i590_print_model_with_imports

agarny commented:
> I can't quite tell who can approve this now that three of us have worked on it? @hsorby @agarny @nickerso ?

I don't know about @hsorby's changes, but mine were purely cosmetic. So, I think it's fine for me to still review this PR.

agarny commented:
> Looks good. I have cleaned up various things though. One thing that I would like to see however is for the `connectionsPersistAfterImporting`, `connectionsAsymmetricWithRepeatedUnits` and `connectionsSymmetricWithMismatchedUnits` to check the serialisation of the model. Right now, we test that this or that component/variable exists, etc. which makes the test rather difficult to read/understand (for me, at least).

Note that I would like to see the above... unless someone can convince that it's not necessary... :)

kerimoyle commented:
> Note that I would like to see the above... unless someone can convince that it's not necessary... :)

@agarny! It's not necessary! (Did that work?) 

I disagree ... I think in general it's better to have nicely isolated tests (ie: covering as little of the API as possible), and I find it hard to read big chunks of raw CellML, when all the details of the test are quite small.  A couple of the other tests here use the parser/printer, but only where that's relevant to what the test is demonstrating.  The majority do not.  What do others think? @hsorby? @nickerso?

hsorby commented:
I am very pro building minimal tests for unit tests.  Only building what is actually needed to exercise the API under test.

kerimoyle committed:
Still a bug here ... see new tests.  Need recursion in forming the import lists.

kerimoyle commented:
Edit: Ignore me.  All good.

~I've found another bug ... so this needs to be reworked :( and could possibly address #652 at the same time as they're related to printing imports.  We need to:~
- ~add some recursion into the printer when forming lists of imports (at present it only checks direct children)~
- ~include imported units in the same forming function (at present only collects components).~

kerimoyle committed:
Revert "Still a bug here ... see new tests.  Need recursion in forming the import lists."

This reverts commit 083787b3517e4946aca6dc5a03ad1aa4020f89a8.

kerimoyle committed:
trigger new build

kerimoyle committed:
Merge branch 'develop' into i590_print_model_with_imports

kerimoyle committed:
Merge branch 'develop' into i590_print_model_with_imports

kerimoyle committed:
Merge branch 'develop' into i590_print_model_with_imports

kerimoyle committed:
Merge branch 'develop' into i590_print_model_with_imports

kerimoyle committed:
Fix merge conflicts

kerimoyle commented:
Booooo ... @hsorby the gremlins are back ... CI is building on f32d900 and the fix in 2fb347d means the tests should pass :(

kerimoyle committed:
Merge branch 'develop' into i590_print_model_with_imports

kerimoyle committed:
Merge branch 'develop' into i590_print_model_with_imports

kerimoyle committed:
Merge branch 'develop' into i590_print_model_with_imports

kerimoyle committed:
Merge branch 'develop' into i590_print_model_with_imports

kerimoyle committed:
Merge branch 'develop' into i590_print_model_with_imports

kerimoyle committed:
Merge branch 'develop' into i590_print_model_with_imports

hsorby commented on tests/test_utils.cpp:
```diff
@@ -60,87 +60,104 @@ void printIssues(const libcellml::LoggerPtr &l, bool headings, bool causes, bool
-        std::cout << "  Units[" << u << "] is '" << model->units(u)->name() << "'"
-                  << std::endl;
-    }
+static const std::string INDENT = "    ";
```
To reduce confusion with later uses of `indent` I think this variable should be renamed to `FIXED_INDENT`.

kerimoyle replied:
f0f249dc

hsorby commented on tests/variable/variable.cpp:
```diff
@@ -1700,6 +1700,115 @@ TEST(Variable, variableInterfaceDontDowngrade)
+    EXPECT_EQ(x2, x1->equivalentVariable(0));
+}
+
+TEST(Variable, connectionsAsymmetricWithRepeatedUnits)
```
I don't understand what this test is doing.  If I remove it I don't fail any tests nor reduce coverage below 100%.

kerimoyle replied:
1) Removing a test kinda means that it's not gonna fail ... ;) 
2) Test coverage should test all the behaviours, not just all the lines of code, no? 100% coverage should be the bare minimum ...
3) It's been so long I'm struggling to remember why they were written.  I suspect it was to do with mismatched units causing an equivalence not to be added, hence the two tests where one (this one) has valid units between the connected variables, and the other where they're not the same (ms and mm).  This PR came through a maze of issues though, so I'll delete the tests as I'm pretty sure the behaviour is covered elsewhere, and the changes to the `Importer` class will overwrite much of this in any case.  8114ded1

hsorby commented on tests/variable/variable.cpp:
```diff
@@ -1700,6 +1700,115 @@ TEST(Variable, variableInterfaceDontDowngrade)
+    EXPECT_EQ(x2, x1->equivalentVariable(0));
+}
+
+TEST(Variable, connectionsSymmetricWithMismatchedUnits)
```
I don't understand what this test is doing.  If I remove it I don't fail any tests nor reduce coverage below 100%.

kerimoyle replied:
8114ded1

kerimoyle committed:
renamed INDENT to FIXED_INDENT

kerimoyle committed:
deleted tests ...

kerimoyle committed:
Merge branch 'develop' into i590_print_model_with_imports

agarny commented:
This PR seems to be broken now.

hsorby commented:
Just minor API change required I think.

hsorby committed:
Fix connectionsPersistAfterImporting text in tests/variable/variable.cpp.

agarny commented on tests/printer/printer.cpp:
```diff
@@ -294,6 +294,40 @@ TEST(Printer, printModelWithStandardUnitsAdded)
+    auto parser = libcellml::Parser::create();
+    auto modelContents = fileContents("importingModelParentComponent.cellml");
+    auto model = parser->parseModel(modelContents);
+    EXPECT_EQ(size_t(0), parser->errorCount());
```
Any particular reason why it's `errorCount()` and not `issueCount()`?

kerimoyle replied:
In general we should be using what we want to be checking.  I think there's even an open issue for exactly that.  As the Issue class becomes more widely used, we'll start to see information items stored in the logger, so it's better to check for errors when you mean to check for errors, and issues when you mean to check for anything.

agarny commented on tests/printer/printer.cpp:
```diff
@@ -294,6 +294,40 @@ TEST(Printer, printModelWithStandardUnitsAdded)
+{
+    auto parser = libcellml::Parser::create();
+    auto modelContents = fileContents("importingModelChildComponent.cellml");
+    auto model = parser->parseModel(modelContents);
```
I guess it's "fine" not to test that the parser didn't generate any errors...?

kerimoyle replied:
We don't check everything everywhere every single time ... so yeah, I reckon it's fine.

agarny commented on tests/resources/importedModelWithCustomUnits1.cellml:
```diff
@@ -0,0 +1,9 @@
+  <component name="importMe">
+    <variable name="x" units="myUnits" interface="public_and_private"/>
+  </component>
+</model>
```
This CellML file is not being used anymore. Delete it.

kerimoyle replied:
47e287cd

agarny commented on tests/resources/importedModelWithCustomUnits2.cellml:
```diff
@@ -0,0 +1,9 @@
+  <component name="importMe">
+    <variable name="x" units="myUnits2" interface="public_and_private"/>
+  </component>
+</model>
```
This CellML file is not being used anymore. Delete it.

kerimoyle replied:
47e287cd

agarny reviewed (changes requested):
Looks good apart from a few minor requests.

kerimoyle committed:
deleted unused test files
